generate-docs.sh

```diff
@@ -0,0 +1,51 @@
+#!/bin/sh
+
+set -e
+
+# Build API documentation, Matlab (.m) -> Markdown (.md)
+#   Beware: you MUST put `quit` at the end; Matlab does not auto-quit.
+#   https://www.mathworks.com/matlabcentral/answers/523194-matlab-script-in-batch-from-unix-command-line
+# TODO: broken.
+#  1. writes to temp/shimming-toolbox/docs instead of docs/
+#  2. incompatible with Matlab R2019a which is what's on our CI machine.
+#matlab -nodisplay -nosplash -r "run('genDoc/generate_doc.m');exit"
+
+# run mkdocs
+# ported from https://github.com/mhausenblas/mkdocs-deploy-gh-pages/blob/master/action.sh
+# with hints from https://github.com/DavidS/jekyll-deploy/blob/master/entrypoint.rb and https://docs.travis-ci.com/user/deployment/pages/
+# maybe it should be its own script?
+
+# install mkdocs
+#export PATH="~/.local/bin":"$PATH"
+#pip install --user mkdocs
+
+# we need to do this in an explicit venv since our build agent
+# isn't fancy enough to spawn fresh containers/VMs for us
+pip install --user virtualenv
+VENV=$(mktemp -d)
+python -m virtualenv "$VENV"
+. "$VENV"/bin/activate
+pip install mkdocs
+
+if [ -n "${GH_PAGES_TOKEN}" ]; then
+    # XXX this is destructive to the repo, but we're assuming this is being run in CI so no matter?
+
+    #git config http."https://github.com/".extraheader "Authorization: token $GH_PAGES_TOKEN" # authenticate over HTTPS with the given token
+    # ^ this doesn't work. github will accept 'token ...' for API calls, but not for git+https://
+    # so instead we have to generate an HTTP basic auth:
+    auth="$(echo -n "x-access-token:$GH_PAGES_TOKEN" | base64 | tr -d '\r' | tr -d '\n')"
+    git config http."https://github.com/".extraheader "Authorization: basic "$auth"" # authenticate over HTTPS with the given token
+
+    git config url."https://github".insteadOf "git@github.com:"  # force HTTPS, not SSH
```

Comment:
Auth is hard.

Computing the base64 in shell like this is unfortunate. I don't like it but it's the best I've got so far.

[Microsoft suggests](https://cloudblogs.microsoft.com/opensource/2019/04/05/publishing-github-pages-from-azure-pipelines/) using
[Github Deploy Keys](https://developer.github.com/v3/guides/managing-deploy-keys/#deploy-keys). but that seems even harder.

https://github.com/DavidS/jekyll-deploy/blob/master/entrypoint.rb and all other Github Actions deploy scripts benefits from the [built in logic](https://github.com/actions/checkout/blob/master/src/git-auth-helper.ts#L53-L66) which does exactly this algorithm before passing the repo off to the CI script.

The Action I ported from, https://github.com/mhausenblas/mkdocs-deploy-gh-pages, thought the thing to do was to replace the URL entirely with one with `x-access-token:$GH_PAGES_TOKEN` in it, and let the HTTP library handle the base64 step. But, not intending to run this on Github's infra, I don't have a reliable way to get the repo URLs besides the ones that come with the repo, so I would rather not try overwriting them.

I tried passing `Authorization: token $GH_PAGES_TOKEN` which should have worked -- it worked for accessing api.github.com: https://github.com/DavidS/jekyll-deploy/issues/7 -- but I couldn't get it to go.

Reply:
It turns out http.extraHeader doesn't even work on our CI machine anyway: its git is too old. So I did it the jekyll-deploy way afterall.